contrib/lib/src/databases.rs

```diff
@@ -0,0 +1,245 @@
+    pub extras: &'a BTreeMap<String, Value>,
+}
+
+/// A custom error type to wrap multiple, disparate error types
+///
+/// This allows for returning the given error from the connection manager
+/// creation methods and the R2D2 connection pool methods.
+#[derive(Debug)]
+pub enum DbError<T> {
+    Custom(T),
+    PoolError(r2d2::Error),
+}
+
+/// The error type for fetching the DatabaseConfig
+#[derive(Debug)]
+pub enum DatabaseConfigError {
+    MissingTable,
+    MalformedConfiguration,
+    MissingConnectionUrl,
+    InvalidConnectionUrl,
+    InvalidPoolSize(i64),
+}
+
+impl<'a> Display for DatabaseConfigError {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match self {
+            DatabaseConfigError::MissingTable => {
+                write!(f, "Database-specific configuration was not found for this database")
+            },
+            DatabaseConfigError::MalformedConfiguration => {
+                write!(f, "The configuration for this database is malformed")
+            }
+            DatabaseConfigError::MissingConnectionUrl => {
+                write!(f, "The connection URL is missing for this database")
+            },
+            DatabaseConfigError::InvalidConnectionUrl => {
+                write!(f, "The specified connection URL is invalid")
+            },
+            DatabaseConfigError::InvalidPoolSize(invalid_size) => {
+                write!(f, "'{}' is not a valid value for `pool_size`", invalid_size)
+            },
+        }
+    }
+}
+
+/// This method retrieves the database configuration from the `Rocket.toml` file
+/// and returns a DatabaseConfig struct.
+///
+/// Example usage:
+/// ```rust,ignore
+/// let config = database_config("my_database", rocket.config());
+/// ```
+pub fn database_config<'a>(
+    name: &str,
+    from: &'a config::Config
+) -> Result<DatabaseConfig<'a>, DatabaseConfigError> {
+    // Find the first `databases` config that's a table with a key of 'name'
+    // equal to `name`.
+    let connection_config = from.get_table("databases")
+        .map_err(|_| DatabaseConfigError::MissingTable)?
+        .get(name)
+        .iter()
+        .filter_map(|config| config.as_table())
+        .next()
+        .ok_or(DatabaseConfigError::MalformedConfiguration)?;
+
+    let maybe_url = connection_config.get("url")
+        .ok_or(DatabaseConfigError::MissingConnectionUrl)?;
+
+    let url = maybe_url.as_str().ok_or(DatabaseConfigError::InvalidConnectionUrl)?;
+
+    let pool_size = connection_config.get("pool_size")
+        .and_then(Value::as_integer)
+        .unwrap_or(from.workers as i64);
+
+    if pool_size < 1 || pool_size > u32::max_value() as i64 {
+        return Err(DatabaseConfigError::InvalidPoolSize(pool_size));
+    }
+
+    Ok(DatabaseConfig { url, pool_size: pool_size as u32, extras: connection_config })
+}
+
+/// This trait can be implemented on Connection types that can be stored in a connection pool.
```

Comment:
This doc needs to be improved. See how `FromRequest` is implemented as an example of where we need to be.